pkg/core/store/badger/extension.go

```diff
@@ -0,0 +1,340 @@
+package badger
+
+import (
+	"context"
+	"time"
+
+	"github.com/fuseml/fuseml-core/pkg/domain"
+	"github.com/timshannon/badgerhold/v3"
+)
+
+// ExtensionStore is a wrapper around a badgerhold.Store that implements the domain.ExtensionStore interface.
+type ExtensionStore struct {
+	store *badgerhold.Store
+}
+
+// NewExtensionStore creates a new ExtensionStore.
+func NewExtensionStore(store *badgerhold.Store) *ExtensionStore {
+	return &ExtensionStore{store: store}
+}
+
+// AddExtension adds a new extension to the store.
+func (es *ExtensionStore) AddExtension(ctx context.Context, extension *domain.Extension) (*domain.Extension, error) {
+	extension.EnsureID(ctx, es)
+	extension.SetCreated(ctx)
+
+	err := es.store.Insert(extension.ID, extension)
+	if err != nil {
+		return nil, domain.NewErrExtensionExists(extension.ID)
+	}
+	return extension, nil
+}
+
+// GetExtension retrieves an extension by its ID.
+func (es *ExtensionStore) GetExtension(ctx context.Context, extensionID string) (*domain.Extension, error) {
+	extension := &domain.Extension{}
+	err := es.store.Get(extensionID, extension)
+	if err != nil {
+		return nil, domain.NewErrExtensionNotFound(extensionID)
+	}
+	return extension, nil
+}
+
+// ListExtensions retrieves all stored extensions.
+func (es *ExtensionStore) ListExtensions(ctx context.Context, query *domain.ExtensionQuery) (result []*domain.Extension) {
+	result = []*domain.Extension{}
+
+	// TODO: Replace with a badgerhold query.
+	if query != nil {
+		if query.ExtensionID != "" {
+			fullExtension, err := es.GetExtension(ctx, query.ExtensionID)
+			if err == nil {
+				matchingExtension := fullExtension.GetExtensionIfMatch(query)
+				if matchingExtension != nil {
+					result = append(result, matchingExtension)
+				}
+			}
+			return
+		}
+
+		allExtensions := []*domain.Extension{}
+		es.store.Find(&allExtensions, nil)
+
+		for _, extension := range allExtensions {
+			matchingExtension := extension.GetExtensionIfMatch(query)
+			if matchingExtension != nil {
+				result = append(result, matchingExtension)
+			}
+		}
+		return
+	}
+
+	es.store.Find(&result, nil)
+	return
+}
+
+// UpdateExtension updates an existing extension.
+func (es *ExtensionStore) UpdateExtension(ctx context.Context, newExtension *domain.Extension) error {
+	extension, err := es.GetExtension(ctx, newExtension.ID)
+	if err != nil {
+		return err
+	}
+	newExtension.Created = extension.Created
+	newExtension.Updated = time.Now()
+
+	for _, newExtService := range newExtension.ListServices() {
+		_, err := extension.GetService(newExtService.ID)
+		if err != nil {
+			// If the service is new, set the creation time
+			newExtService.SetCreated(newExtension.Updated)
+		}
+	}
```

Comment:
I could be wrong, but I thought there was a domain.Extension method that would do this for you, so you wouldn't have to duplicate it here.